### PR TITLE
Port to nanopb

### PR DESCRIPTION
**Pull request by ezyang**
General strategy:
- nanopb is statically linked into PyTorch.  It must be built
  with -fPIC.
- Generated nanopb files for toffee.proto are checked into
  our repo.
- Because nanopb generated protobufs are C only, we wrote a
  wrapper around it to give a Google C++ style interface.
  More on this shortly.

How does the wrapper work?
- It's called "micropb" becaues it is less small than nanopb :)
- nanopb requires all variable-length fields to be written out
  using a "callbacks" mechanism.
- We wrote pre-canned callbacks for all of the types ToffeeIR
  writes out and lists; these are micropb_callback and
  micropb_callback_list.  These operate simply by dynamically
  allocating and storing the data to be written out in
  data (this defeats the purpose of the callback mechanism,
  but it's easy to implement)
- MicroProto class handles memory management of these dynamically
  allocated pieces of data.  This is done by a type-erased
  unique_ptr<void> with runtime deleter, so it's not required
  to write owning references as member fields of MicroProto.
- Finally some boilerplate to actually implement the wrapper
  classes

Testing strategy:
- Take the serialized protobuf from nanopb, parse it again
  with ToffeeIR and print it.  Worked with all of test_jit.py!

### CODE REVIEW

**Comment by zdevito** [comment]
I really like the overall idea of preserving the normal protobuf API while using nanopb since it allows us to swap later if we need too. I don't get the memory management stuff going on here, though.   Why can't NodeProto, GraphProto, etc. own the memory for themselves and all their non-fixed-sized members, and then let the callbacks work off of that object hierarchy?

**Comment by ezyang** [comment]
I originally had more lofty goals for the memory management stuff, wherein it would not be necessary to declare any fields at all in the subclasses. But I ended up stashing fields for all the vector things, because if a user subsequently calls add_input I need a vector to push into, and it seemed dirty to static_cast it out of the arg pointer in the actual protobuf. At the cost of more boilerplate, I can ditch the fancy memory management entirely; this would probably be wise.

File: torch/csrc/toffee/export.cpp [hunk 1]
```diff
@@ -24,20 +25,24 @@ std::string ExportGraph(std::shared_ptr<Graph>& g) {
...
   }
+
   for (auto node : g->nodes()) {
     if (node->kind() == kSelect) {
       // No select nodes in ToffeeIR: instead we make use
       // of the select invariant
       continue;
     }
     auto generic_node = [&]() {
+      // p_g_node.emplace_back();
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: ??

**Review by apaszke — APPROVED** [review]
Looks good. Would be nice to completely remove dependency on google/protobuf

**Comment by ezyang** [comment]
OK, this is ready to land, but I am stalling it until #112 lands.

**Pull request closed without merging**